### PR TITLE
Pev  fix lsh bad any cast

### PR DESCRIPTION
**Pull request by pemmanuelviel**
closes #17628

<cut/>

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under OpenCV (BSD) License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

```
allow_multiple_commits=1
```

### CODE REVIEW

File: modules/flann/include/opencv2/flann/lsh_index.h [hunk 1]
```diff
@@ -58,9 +58,9 @@ namespace cvflann
 
 struct LshIndexParams : public IndexParams
 {
-    LshIndexParams(unsigned int table_number = 12, unsigned int key_size = 20, unsigned int multi_probe_level = 2)
+    LshIndexParams(int table_number = 12, int key_size = 20, int multi_probe_level = 2)
```
[the remark this answers is not in the document]
Comment (pemmanuelviel) [reply modules/flann/include/opencv2/flann/lsh_index.h]: Good catch, thanks.
I've made the change to see if this passes the tests now.
Adding the test will be my next task.

File: modules/flann/include/opencv2/flann/lsh_index.h [hunk 2]
```diff
@@ -378,11 +378,11 @@ class LshIndex : public NNIndex<Distance>
...
 
     /** table number */
-    unsigned int table_number_;
+    int table_number_;
     /** key size */
-    unsigned int key_size_;
+    int key_size_;
     /** How far should we look for neighbors in multi-probe LSH */
-    unsigned int multi_probe_level_;
+    int multi_probe_level_;
```
Comment (alalek) [inline modules/flann/include/opencv2/flann/lsh_index.h:385]: Please revert these changes too

Comment (pemmanuelviel) [reply modules/flann/include/opencv2/flann/lsh_index.h:385]: I've squashed all these commits in a new branch pev--fix-bad-any-cast-in-lsh based on 3.4, including a test.
So we don't change any unsigned int for int type for table_number_, key_size_ and multi_probe_level_.
However, as
- in all the others types of IndexParams the types are int and not unsigned int
- and the version of these three parameters in LshIndexParams of miniflann is int and not unsigned int,
I would suggest for consistency to switch to int in master branch, what is actually done by the "pev-fix-lsh-bad-any-cast" (one dash and not two after pev to easily identify between master based branched and 3.4 based ones)

**Comment by pemmanuelviel** [comment]
@alalek your forced push let the members declared as int instead of their original unsigned int type (the revert mentioned earlier).
I've pushed another commit (b0315a74a6a91b1446ee56a47b67fe6e675b95ec) to change that on the branch, but it doesn't seem to appear in the PR.

**Comment by alalek** [comment]
@pemmanuelviel According to statement from #17684:

> in all the others types of IndexParams the types are int an not unsigned int

lets change (keep from original patch) "int" type for fields.

I believe we are ready to merge this PR in the current state.

**Comment by pemmanuelviel** [comment]
@alalek I was not sure you wanted to keep the change.
So yes, we can merge this PR without my latest commit b0315a74a6a91b1446ee56a47b67fe6e675b95ec.

Talking about consistency between uint and int, what do you think about my suggestion to merge PR 17628 on master to change also the LshIndexParams arguments type so they become int
- like for others xxxIndexParams,
- and are identical to the ones of the LshIndexParams in miniflann.hpp (l.121) and miniflann.cpp (l.276)?

**Comment by alalek** [comment]
PR on master (#17628) will be closed after merging this PR and after [regular](https://github.com/opencv/opencv/wiki/Branches) merge of fixes from 3.4 to master branch.

Which type is currently used in upstream flann project for smilar fields in `xxxIndexParams`?

**Comment by pemmanuelviel** [comment]
All of them, except LshIndexParams, are using signed integers in their respective header files:
- CompositeIndexParams(int trees = 4, int branching = 32, int iterations = 11,
- HierarchicalClusteringIndexParams(int branching = 32,
- KDTreeIndexParams(int trees = 4)
- KDTreeSingleIndexParams(int leaf_max_size = 10, bool reorder = true, int dim = -1)
- KMeansIndexParams(int branching = 32, int iterations = 11,
And for their mapping in miniflann, even LshIndexParams use signed integers.

**Comment by alalek** [comment]
@pemmanuelviel Thank you for information! Actually my question was about `int table_number_` fields from `LshIndex`.

**Comment by pemmanuelviel** [comment]
@alalek Each type of tree has its own set of parameters, with some equivalences for some of them.
Thus "table_number" is the equivalent for LSH of "trees" for KDTree and "branching"  for KMeans and HierachicalClustering.

**Comment by alalek** [comment]
Third commit is omitted due to GitHub todays incident.
PR is merged with commit cdac7c7bec3d1983b3a635a49d1b69cf07d0bdb7

[https://github.com/opencv/opencv/pull/17640/commits/b0315a74a6a91b1446ee56a47b67fe6e675b95ec ](https://github.com/opencv/opencv/pull/17640/commits/b0315a74a6a91b1446ee56a47b67fe6e675b95ec)

> This commit does not belong to any branch on this repository.

Still broken.

PR is not marked as merged automatically due GitHub malfunction. Closing...
